Fix: tools: crm_mon --daemonize should update when disconnected

nrwahl2 opened the pull request:
With `crm_mon --daemonize`, the output will continue to show the last known status after the cluster is stopped on the local node. It should go back to something like `"Cluster is not available"`.

Likewise, messages like `"Reconnecting..."` should not go to the daemonized output. The output file (or external handler) should receive only the cluster status. So we print those messages only if `output_format` is `mon_output_console` (or if we're in one-shot mode, where applicable). Normally those messages don't get printed to the daemonized output anyway because we don't flush the buffer, but theoretically they could get printed if the buffer fills up.

It's questionable whether we want to print `"Cluster is not available"` when we're in interactive mode or only in daemonized mode. For now, we'll do daemonized-only. Interactive (console) mode already gets the `"Connection to cluster daemons terminated"` message, which is quickly replaced by `"Reconnecting..."`.

External agents are notified via traps rather than via output, so we can ignore them.

Closes T15

nrwahl2 commented:
The first commit may not be comprehensive -- for example, "waiting for start" type messages may still appear during startup. It's an improvement and should take care of most situations after we're already up and running.

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -2009,10 +2010,30 @@ mon_refresh_display(gpointer user_data)
+        exit_code = CRM_EX_DISCONNECT;
+
+        // Print an unavailable message for daemonized crm_mon
+        switch (output_format) {
```
We **could** be create formatter functions that simply print "Cluster is not available" in the right way for each format, and call `out->message()` here. That would be a bit more involved than this approach (and would require adding an attribute or element to the XML schema), but it's certainly doable.

kgaillot replied:
Definitely, we should never switch on format to choose what display

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -2070,7 +2091,7 @@ refresh_after_event(gboolean data_updated, gboolean enforce)
     }
 
-    if (reconnect_timer > 0) {
+    if (!enforce && (reconnect_timer > 0)) {
```
This gets us around the early return so that we set the `refresh_trigger`, calling `mon_refresh_display()`.

kgaillot replied:
the idea is if we're about to reconnect, there's no point in refreshing yet. but if we need to display some state info in the meantime, something else might be appropriate

kgaillot commented on tools/crm_mon.c:
```diff
@@ -655,7 +655,10 @@ reconnect_after_timeout(gpointer data)
+    if (output_format == mon_output_console) {
+        out->info(out, "Reconnecting...");
+    }
+
```
We should avoid checking the format directly whenever possible. We should use messages for all these. However it feels like there's a logical affinity here, so it might be better to create a new output method to use with all of them and make it a no-op for everything but console (the log formatter could maybe log it at debug level).

Maybe something like ->transient()

nrwahl2 replied:
Do we need to make a separate output type for console in order to do that? Right now there isn't one. It seems like it would just be a duplicate of `output_text.c` with names changed, other than implementing `out->transient()`.

nrwahl2 replied:
On second thought, by the time we were to implement an `output_console` type, that would defeat the purpose of having `transient()`. `out->info()` would effectively be transient for the console.

kgaillot replied:
There is a console format -- it is implemented solely by crm_mon, in crm_mon_curses.c. Which reminds me, that's one of the reasons it's always good to have a default implementation for custom messages, because the console can use that if there's no explicit implementation in crm_mon_curses.c.

kgaillot commented on tools/crm_mon.c:
```diff
@@ -952,8 +957,9 @@ pacemakerd_status(void)
-                        (output_format == mon_output_plain)) {
+                    if (options.one_shot
+                        || (output_format == mon_output_console)) {
+
```
yeah this block is a good example, there shouldn't be an if/else here, just calling appropriate message functions (in this case probably a single message function that takes `state`)

that would also let us indicate this in xml format somehow

nrwahl2 replied:
If we even want to indicate it in non-console formats, sure. It depends on whether we think these should be transient messages like "Reconnecting...". If so, then they'd be console-only (plus `log` at the debug level) via `out->transient()`, and we wouldn't need messages configured specifically for this task.

I'm on the fence. I lean toward making them all transient, as I don't see much of a reward to be gained from calling a message here, and in general we should progress to a new state quickly. If we do go with messages, maybe we can repurpose the `pacemakerd-health` message functions.

Especially since so far, I haven't been able to get into these states to get any of these to actually print from `pacemakerd_status()` using any format or `crm_mon` run option. So maybe it's pretty rare that they would even appear?

I've got a draft of a message approach ready to go, but for the reasons discussed above, it doesn't seem to make a difference.

nrwahl2 replied:
We can do messages for the "Cluster is not available" message, regardless of what we decide for `pacemakerd_status()`.

nrwahl2 replied:
> For console, the ideal is that if everything proceeded in slow motion, the user would see a succession of messages indicating the start-up/shutdown progress.

That makes sense. That doesn't happen currently.

I added a sleep after each `pacemakerd_state` assignment and before the first assignment. In console mode, I get nothing but `"Waiting until cluster is available on this node ..."` when we're waiting for the first connection, and `"Connection to the cluster-daemons terminated"` followed by `"Reconnecting ..."` when we lost our connection (e.g., due to stopped cluster) and we're waiting to reconnect. We don't see any of the messages from `pacemakerd_status()`. It's a similar story in one-shot mode.

It might be due to the way resets are ordered relative to the `pacemakerd_status()` messages. I'll look more closely soon. Just saying we don't have to worry specifically about preserving the current display of those messages, since they don't display.

Originally I hoped to just use `out->transient()` calls. If we want to show these states in XML output too, then transient's not gonna work. It might not end up making sense to add the `pcmk__output_t:transient()` method in the first place, depending on what we decide to do with the `Waiting until...` and `Reconnecting...` messages (the first intended use case for `transient`).

kgaillot replied:
> > It could even be a blanket "cluster status not yet available" message or something like that.
> 
> We currently do get a generic message like that, for one-shot or console mode. For one-shot, you get `"crm_mon: Connection to cluster failed: Resource temporarily unavailable"`. For console, you get `"Waiting until cluster is available on this node ..."` if you haven't established a connection yet and `"Reconnecting ..."` if you have. Daemonized mode is the only case where you don't get a generic "cluster not available" type of message yet.

That's not bad, though it would be ideal for daemonized HTML/CGI to say something useful.

> 
> > For console, the ideal is that if everything proceeded in slow motion, the user would see a succession of messages indicating the start-up/shutdown progress.
> 
> That makes sense. I'm not sure that happens currently. It looks like even if you can catch the usually-ephemeral intermediate `pacemakerd` states, the `Waiting until cluster is available on this node ...` message or the `Reconnecting ...` message will clear the screen and overwrite the state before anyone observes it. One-shot mode might behave better if you can time it right, not sure.

Bah. Maybe pacemakerd_status() could give back a string that could be tacked on to "Waiting until ..." e.g. "(Pacemaker daemons starting)".

> 
> Adding sleeps in `pacemakerd` turns out not to work because it causes `pcmk_ipc_connect()` calls to hang in `pacemakerd_status()`. So while we're sleeping, we can't get the `pacemakerd` state. Delaying the state updates by doing it after some interval through the mainloop instead of synchronously would probably get around that.

Doh, of course. That may be more effort than this is worth.


> Gotta decide what to do with the `"Waiting until..."` and `"Reconnecting..."` messages if we have a specific `pacemakerd` state to display (considering the `clear()` calls and such)... and other than that, it might be a straightforward addition of a new message for each output format. We might be able to reuse the `pacemakerd-health` message for this, but it would require some hackiness:
> 
>     * text mode prints fields we don't want unless we're in quiet mode
> 
>     * console mode clears the screen which would mess with the display of transient messages like `"Reconnecting"`
> 
> 
> So a semi-redundant new message may be more appropriate, we'll see
> 
> Also thinking we could merge the `pacemakerd_status()` definitions from `crm_mon.c` and `pcmk_status.c` in some way

I don't remember why those are separate, I think it had something to do with 1-shot vs not

nrwahl2 replied:
> I don't remember why those are separate, I think it had something to do with 1-shot vs not

The one in `crm_mon.c` was added first. The one in `pcmk_status.c` was added later, but at a glance they seem to be doing the same thing with minor differences.
* `pcmk_status.c` doesn't do the conditional printing based on output format.
* `pcmk_status.c` uses `EAGAIN` instead of `ENOTCONN` as its default return code.
* `crm_mon.c` uses `dispatch_poll` while `pcmk_status.c` uses `dispatch_sync`.
* `crm_mon.c` returns `pcmk_rc_ok` early if `!use_cib_native`.
* `pcmk_status.c` prints a "Could not connect" message if `pcmk_ipc_connect()` fails

Might be able to make them both happy with a common internal function

kgaillot commented on tools/crm_mon.c:
```diff
@@ -705,7 +705,7 @@ mon_cib_connection_destroy(gpointer user_data)
                                         reconnect_after_timeout, NULL);
     }
-    return;
+    refresh_after_event(FALSE, TRUE);
```
might as well add this situation to the refresh_after_event() doxygen

nrwahl2 replied:
That lists the conditions when the screen is actually redrawn, assuming that `refresh_after_event()` has already been called. It doesn't list the call conditions

nrwahl2 replied:
Wouldn't hurt to add the "`enforce` is true" condition regardless

kgaillot replied:
ah

kgaillot commented on tools/crm_mon.c:
```diff
@@ -2009,10 +2010,30 @@ mon_refresh_display(gpointer user_data)
-                                     show, show_opts, options.only_node,
-                                     options.only_rsc, options.neg_location_prefix,
-                                     output_format == mon_output_monitor);
+    if ((cib != NULL) && (cib->state != cib_disconnected)) {
```
what happens now when this isn't true? i thought that maybe we would still try to show fence history and pacemakerd status

nrwahl2 replied:
It doesn't care about the CIB state now. `pcmk__output_cluster_status()` uses `current_cib`. As far as my testing went, `current_cib` stayed populated. Only `cib_connect()` and `pacemakerd_status()` query the CIB and alter `current_cib`.

I also wasn't able to hit `mon_refresh_display()` after a CIB disconnection until I added this patch. Right now, `mon_refresh_display()` only gets called **before** `mon_cib_connection_destroy()`.

If `current_cib` were `NULL`, presumably yes something down the call chain from `pcmk__output_cluster_status()` would end up printing the things we do have access to (fence history, etc.).

Not to say with certainty that anything can't happen... it gets tough to follow.

kgaillot replied:
Test the original code to see if the other status is still shown correctly (running pacemaker-fenced in standalone mode should be sufficient for this purpose). If not, then we can forget about it, otherwise we'll need to do this in pcmk__output_cluster_status() and the "cluster-status" message.

nrwahl2 replied:
If `crm_mon` can't connect to `pacemakerd` (e.g., because the fencer is running in standalone mode), it doesn't show anything except
```
Could not connect to pacemakerd: Connection refused
crm_mon: Connection to cluster failed: Connection refused
```
in one-shot mode, or `Waiting until cluster is available on this node ...` in console mode. And nothing at all in daemonized mode.

If I run `pacemakerd`, that's going to start all the subdaemons. If I then kill `pacemaker-based`, `pacemakerd` restarts `pacemaker-based`, `pacemaker-fenced`, and `pacemaker-controld`. In the brief interim period when daemons are being restarted, the node state is updated to `OFFLINE`, the resource state disappears, and the current DC disappears.

kgaillot replied:
OK, this should be fine then

nrwahl2 commented:
pushed from the wrong local branch...

kgaillot reviewed (commented):
We've improved the crm_mon code drastically over the past few years, but it's still difficult to follow :-/

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -683,7 +746,9 @@ reconnect_after_timeout(gpointer data)
 {
-    out->info(out, "Connection to the cluster-daemons terminated");
+    if (!options.daemonize) {
+        out->info(out, "Connection to the cluster-daemons terminated");
```
Right -- the idea there is to print that message in console mode but not in daemonized text mode.

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -683,7 +746,9 @@ reconnect_after_timeout(gpointer data)
+    }
 
     if (refresh_timer != NULL) {
         /* we'll trigger a refresh after reconnect */
```
True, that should be removed or revised. In this PR's current form, a minor tweak will make this work without triggering an immediate refresh, relying just on the timer. IMO it's better to immediately update the output file or console in response to a *known* change though.

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -705,7 +770,7 @@ mon_cib_connection_destroy(gpointer user_data)
                                         reconnect_after_timeout, NULL);
     }
-    return;
+    refresh_after_event(FALSE, TRUE);
```
Also true. When we added the `enforce` option a while back, it added some untapped opportunities.

nrwahl2 commented:
Depending on how much effort it is, before having this merged, I may try to make the console and daemonized mode share more code with one-shot mode since we're doing that work in #2902. That would entail, at minimum, using `pcmk__pacemakerd_status()`. It could also involve sharing a `cib_connect()` and `fencing_connect()` implementation and maybe incorporating `pcmk__status()`... I haven't given it any detailed thought since starting on #2902.

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -918,97 +927,101 @@ pacemakerd_status(void)
                 out->info(out,
-                    "Running on remote-node waiting to be connected by cluster ...");
+                          "Running on a remote node, waiting to connect to "
+                          "the cluster...");
```
At this point, we don't know anything about whether the Pacemaker Remote node is connected to the cluster. We only know that `pacemaker-remoted` is running on the local host. `crm_mon` has yet to try to connect to the fencer or CIB.

"waiting to connect to the cluster" could suggest that this particular client is not yet connected to the CIB. However, it's probably better to leave it at "Running on a Pacemaker Remote node." That way we avoid any misleading statements.

kgaillot replied:
Not for this PR, but just thinking out loud, I wonder whether pacemaker-remoted should support a ping that returns whether it's connected or not.

nrwahl2 replied:
It sounds like a good idea, at least

nrwahl2 commented:
Updated minimally to address review. There's still room for improvement in both the output and the implementation, but this gets the job done.

I'll be filing another PR with some other refactors as a start

nrwahl2 commented:
The CI failures were RPM issues on three hosts. Two of them were OpenSUSE hosts that timed out when trying to access the libqb100 RPM. The other was CentOS 9 s390x: "  - nothing provides libqb(s390-64) = 2.0.6-1.el9.next needed by libqb-devel-2.0.6-1.el9.next.s390x"

kgaillot commented on tools/crm_mon.c:
```diff
@@ -101,6 +101,68 @@ static pcmk__supported_format_t formats[] = {
+        out->err(out, "%s (%s)", msg, description);
+    } else {
+        out->err(out, "%s", msg);
+    }
```
I'm thinking we don't always want this as an error -- aren't we now calling this before the first connection?

nrwahl2 replied:
Almost certainly better to use the pacemakerd-health message. It's already on my to-do list to do that via `pcmk__pacemakerd_status()`. Wouldn't hurt to do it via pacemakerd-health now, given that we improved that since this PR was first filed.

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -910,6 +910,15 @@ pacemakerd_event_cb(pcmk_ipc_api_t *pacemakerd_api,
+ * \internal
+ * \brief Get the status of \p pacemakerd
+ *
+ * If we need a new CIB connection, check whether \p pacemakerd is in a healthy
```
s/new/new live/? The other idea is that if we already have a connection, we don't need to do anything in this function.

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -918,97 +927,72 @@ pacemakerd_status(void)
+            if ((output_format == mon_output_console)
+                || (output_format == mon_output_plain)) {
+
+                out->info(out, "%s", pcmk__pcmkd_state_enum2friendly(state));
```
Right. I was thinking ahead to making the output closer to the changes we made for one-shot -- we always show the pacemakerd status there. (As a side note, one part of the other changes I've been working on is to make displaying pacemakerd status toggleable like the other sections.)

I'm not tied to changing which states are displayed right now.

kgaillot replied:
It just feels odd to output "Pacemaker is running" before showing the actual status

Maybe the pacemakerd status should be in the summary, maybe like "Stack: corosync (Pacemaker is running)"

nrwahl2 replied:
> It just feels odd to output "Pacemaker is running" before showing the actual status

I agree with that. It's been painful trying to balance intuitive output in all three execution types, with not having a bunch of spaghetti code doing special handling for different states. Showing the `pacemakerd` status unconditionally seems nicer in XML format (where we just have a `<pacemakerd>` element) compared to text/console, FWIW.

I'd also like to eventually have the console/daemonized output as close to the one-shot output as possible (minus the "Waiting for..."-type messages in console). That might involve dropping the "Pacemaker is running" display from the text one-shot output, idk

> Maybe the pacemakerd status should be in the summary, maybe like "Stack: corosync (Pacemaker is running)"

That's definitely something I've been thinking about but I haven't really liked where those thoughts have led. It gets kind of messy implementation-wise. Not that any of this is super clean.

In most `pacemakerd` states, the CIB isn't available, and all the rest of the info in the `cluster-status` message (of which `cluster-summary` is a part) comes from the CIB. Whether it goes in the stack line or on its own line, it necessitates more special handling in `pcmk__output_cluster_status()` and `cluster-status` for when that's the only data available.

If we start using `pcmk__pacemakerd_status()` here and we don't output unconditionally from there, then public API for `pcmk_pacemakerd_status()` would have to either call the `pacemakerd-health` message directly or call something like `pcmk__output_cluster_status()` with the only output flag set for `pacemakerd` status.

I feel like there were other a couple other questionable consequences but they're not coming to mind.

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -648,15 +648,22 @@ static GOptionEntry deprecated_entries[] = {
+    if (desc != NULL) {
+        out->info(out, "%s", desc);
+    }
     out->info(out, "Reconnecting...");
```
Good point -- although it will show "Cluster is not available" (and for console mode, "Connection to the cluster lost") before this. I think I grabbed the status first because in the third commit we do `"Reconnecting (<pcmkd_status>)"`. I'll rethink that

nrwahl2 commented:
still working out some schema issues

kgaillot commented on lib/cib/cib_utils.c:
```diff
@@ -771,9 +771,6 @@ cib__signon_query(pcmk__output_t *out, cib_t **cib, xmlNode **cib_object)
 
-    if (*cib_object == NULL) {
-        return pcmk_rc_no_input;
-    }
```
I think the real intention here was:

    if ((rc == pcmk_rc_ok) && (*cib_object == NULL)) {

I doubt that's possible in practice but it seems like a reasonable failsafe

nrwahl2 replied:
Yeah I think that's better

nrwahl2 commented on tools/crm_mon.c:
```diff
@@ -1472,22 +1481,24 @@ main(int argc, char **argv)
+            clear();
+            refresh();
+        }
+#endif
```
Moving this block to the top of the do-while loop is an accidental leftover. It gets undone in the final commit. I'll undo this change before we merge anything.

kgaillot commented on tools/crm_mon.c:
```diff
@@ -852,149 +848,54 @@ set_fencing_options(int level)
+            case pcmk_pacemakerd_state_remote:
+                /* Fencer and CIB may still be available while shutting down or
+                 * running on a Pacemaker Remote node
+                 */
```
I do think it's useful to indicate if the cluster is shutting down

nrwahl2 replied:
Sure, easy change.

nrwahl2 replied:
Heh, I was wrong about this :)

Due to the mess of when we can/can't call `out->reset()` and `out->finish()` thanks to the possibility of curses output, yesterday's versoin of this PR was never really showing the `pacemakerd` status at all. It was being called, but the output was being reset or cleared before the buffer was flushed.

I ended up doing something you suggested earlier, that I had decided against at the time: add the `pacemakerd` status to the cluster summary if we have a native CIB connection and get that far (and still output the `pacemakerd` status by itself if we don't get that far). It's the cleanest option I found to implement (the other refactors helped), and the resulting output looks pretty nice IMO. It shows the Pacemaker status even if it's "running," but it doesn't look out of place since it's a summary list item.

```
Cluster Summary:
  * Stack: corosync
  * Pacemaker status: Pacemaker is running
  * Current DC:	node1 (version 2.1.5-0b3656e85) - partition with quorum
  * Last updated: Fri Nov  4 17:35:44 2022
  * Last change:  Fri Nov  4 17:06:38 2022 by root via cibadmin on node1
  * 1 node configured
  * 2 resource instances configured
```

kgaillot replied:
What do you think of something like

    * Stack: corosync (Pacemaker is running)

or 

    * Stack: Pacemaker is running with corosync as cluster layer

to conserve screen space? It fits the concept of "stack" which is Pacemaker layered on top of Corosync. We could tweak the pacemaker messages to fit the usage.

kgaillot replied:
This would also avoid needing new display options since the stack option would control it. The XML for \<stack\> could include the pacemakerd status as a separate attribute.

kgaillot commented on tools/crm_mon.c:
```diff
@@ -672,7 +672,7 @@ reconnect_after_timeout(gpointer data)
 mon_cib_connection_destroy(gpointer user_data)
 {
-    out->info(out, "Connection to the cluster-daemons terminated");
+    out->transient(out, "\nConnection to the cluster lost");
```
Why a newline?

nrwahl2 replied:
Just because the string of ellipses (of varying length) before the first "Connection lost" message always bugged me:
```
Cluster Summary:
  * Stack: corosync
  * Current DC:	node1 (version 2.1.5-0b3656e85) - partition with quorum
  * Last updated: Thu Nov  3 17:02:27 2022
  * Last change:  Thu Nov  3 15:59:10 2022 by hacluster via crmd on node1
  * 1 node configured
  * 2 resource instances configured (1 DISABLED)

Node List:
  * Online: [ node1 ]

Active Resources:
  * dummy	(ocf:pacemaker:Dummy):   Started node1
......Connection to the cluster-daemons terminated
Connection to the cluster-daemons terminated
```

I'm not tied to the newline though. Other options besides the newline:
* Leave it as-is
* Modify the curses error and info/transient functions so that they move the cursor to the beginning of the current line before printing. Not sure if that would have any unintended consequences, but it definitely looks the nicest in this scenario:
```
Cluster Summary:
...
Active Resources:
  * dummy	(ocf:pacemaker:Dummy):   Started node1
Connection to the cluster lost
Connection to the cluster lost
```

kgaillot replied:
I definitely think the handling should be in the implementation rather than the message string. The idea is that the message string should be identical across all formats and not have format-specific control characters.

Moving to the beginning of the line should be fine for error at least (clearing the line would also be a good idea in case the existing content is longer than the error message). For info/transient it would probably be fine too. I suppose the ideal would be to check the line position and output a newline if not at the beginning, but that might not be worth the hassle.

nrwahl2 replied:
I'm going to just remove the newline from the commit for now. We can revisit this later.

> that might not be worth the hassle.

For curses, doing something based on the current position is no more difficult than moving to the beginning of the line. Which is to say, not very. It's probably doable in other formats as well (something like get the character at `ftell(fd) - 1` and act if that character is not a newline? Haven't done much with streams).

There may or may not be existing places where we build part of a string piece-by-piece based on conditions, and then finish it with `out->info(out, "the rest")`. Any new approach that adds a newline or overwrites part of a line would break that.

kgaillot replied:
There shouldn't be, we rejected that possibility when designing the interface. We use the usual techniques to build a string then pass the whole string to the output object function. That's the only way to abstract e.g. text vs xml needing a newline or not. We considered separate methods for build-up-a-line and end-a-line but there are already functions for that purpose, and not all formats are line-oriented.

Just dropping the newline is fine for this PR

kgaillot commented on tools/crm_mon.c:
```diff
@@ -1472,22 +1481,24 @@ main(int argc, char **argv)
-        }
 
+            // Give some time to view all output even if we won't retry
+            pcmk__sleep_ms(options.reconnect_ms);
```
Maybe curses' transient() should sleep like err()?

nrwahl2 replied:
I feel like the cons outweigh the pros, since we don't want to sleep in the common case. That would mean sleeping after "Connecting to cluster...", "Reconnecting...", etc., before we even try to do any work.

In this case, we want to sleep for `reconnect_ms` anyway before retrying if we hit `ENOTCONN` or `ECONNREFUSED`. If we somehow hit a different error (probably rare and probably indicates a bigger issue), then we give up. Putting the sleep outside the if-block just adds a one-time wait of `reconnect_ms` before we clean up and exit.

kgaillot reviewed (commented):
looks good

nrwahl2 commented:
Updated to address review, which was more involved than expected (see https://github.com/ClusterLabs/pacemaker/pull/2868#discussion_r1014536512). This is already way too big, so if it gets any bigger, it'll be split up.
* Updated `cib__signon_query()` to return `pcmk_rc_no_input` if `(rc == pcmk_rc_ok) && (*cib_object == NULL)`.
* Added a commit at the beginning to fix the pcmkd state display for XML format (should use the short display instead of friendly).
* Added two commits to include `pacemakerd` status in the `cluster-summary` message for native CIB connections. It's excluded by default for file and remote CIB connections (if you include it, it will always say "Unknown"). It's also specifically excluded when `pcmk_simulate` calls `cluster-summary`.

Not addressed yet:
* Newline before "Connection to the cluster lost". Need to figure out whether to leave that as-it-was, add the newline, or modify the curses error/info functions to start at the beginning of the line.

nrwahl2 commented:
Updated to:
* remove the newline from the "Connection to cluster lost" transient message
* move the `pacemakerd` state into the stack section:
```
    <stack type="corosync" pacemakerd-state="running"/>
    <stack type="corosync" pacemakerd-state="shutting_down"/>

  * Stack: corosync (Pacemaker is running)
  * Stack: corosync (Pacemaker daemons are shutting down)
```
* Remove the pacemakerd-health messages from `crm_mon.c` (the `crm-mon-disconnected` message covers this already for the non-one-shot case)
